commandapi-core/src/main/java/dev/jorel/commandapi/AbstractCommandAPICommand.java

```diff
@@ -324,12 +325,29 @@ private List<Argument[]> getArgumentsToRegister(Argument[] argumentsArray) {
 		if (firstOptionalArgumentIndex != -1) {
 			for (int i = 0; i <= argumentsArray.length; i++) {
 				if (i >= firstOptionalArgumentIndex) {
-					Argument[] arguments = (Argument[]) new AbstractArgument[i];
-					System.arraycopy(argumentsArray, 0, arguments, 0, i);
-					argumentsToRegister.add(arguments);
+					List<Argument> arguments = new ArrayList<>();
+					Argument[] argumentsWithoutCombined = (Argument[]) new AbstractArgument[i];
+					System.arraycopy(argumentsArray, 0, argumentsWithoutCombined, 0, i);
+					for (Argument argument : argumentsWithoutCombined) {
+						arguments.addAll(unpackCombinedArguments(argument));
+					}
```

Comment:
This method unpacks arguments early in the list many times when it only needs to unpack them once. To explain what I mean, consider these arguments:
```
.withArguments(
	new StringArgument("A"),
	new StringArgument("B").withPermission("permissionB").combineWith(
		// Using argument combining to share permissions
		// I think this is something someone might try, and it isn't stopped with an error
		// See a different comment to see why I think this might be useless though
		new StringArgument("C")
	)
).withOptionalArguments(
	new StringArgument("D").combineWith(
		new StringArgument("E")
	), 
	new StringArgument("F"), 
	new StringArgument("G")
)
```
After figuring out that this doesn't throw an `OptionalArgumentException` and dose include optional arguments, the second loop runs like this:
```
i = 2
new arguments
argumentsWithoutCombined = {"A", "B"}
    unpack "A" -> ["A"]
    unpack "B" -> ["B", "C"]
arguments = ["A", "B", "C"] -> register

i = 3
new arguments
argumentsWithoutCombined = {"A", "B", "D"}
    unpack "A" -> ["A"]
    unpack "B" -> ["B", "C"]
    unpack "D" -> ["D", "E"]
arguments = ["A", "B", "C", "D", "E"] -> register

i = 4
new arguments
argumentsWithoutCombined = {"A", "B", "D", "F"}
    unpack "A" -> ["A"]
    unpack "B" -> ["B", "C"]
    unpack "D" -> ["D", "E"]
    unpack "F" -> ["F"]
arguments ["A", "B", "C", "D", "E", "F"] -> register

i = 5
new arguments
argumentsWithoutCombined = {"A", "B", "D", "F", "G"}
    unpack "A" -> ["A"]
    unpack "B" -> ["B", "C"]
    unpack "D" -> ["D", "E"]
    unpack "F" -> ["F"]
    unpack "G" -> ["G"]
arguments ["A", "B", "C", "D", "E", "F", "G"] -> register
```
As you can tell, there is a lot of repeated calculations here that overall give this part of the method an `O(n^2)` time complexity. However, since unpacking an argument always gives the same result, this method can actually have an `O(n)` time complexity, like so:
```
new arguments

// handling required arguments
// If something prevented required arguments from having arguments 
// combined to them, these unpacks are also unnecessary. If that is 
// possible (eg. for permission sharing, which is questionably useful), 
// then unpacking here is necessary and expected.
unpack A -> ["A"]
arguments = ["A"]

unpack B -> ["B", "C"]
arguments = ["A", "B", "C"]

// end of required arguments
arguments = ["A", "B", "C"] -> register

// handling optional arguments
unpack "D" -> ["D", "E"]
arguments = ["A", "B", "C", "D", "E"] -> register

unpack "F" -> ["F"]
arguments = ["A", "B", "C", "D", "E", "F"] -> register

unpack "G" -> ["G"]
arguments = ["A", "B", "C", "D", "E", "F", "G"] -> register
```
If you look at when `arguments` get registered, you can see this does the same thing but in many fewer steps. Implementing this `O(n)` algorithm is an exercise left to the viewer :). The implementation I suggested [here](https://github.com/JorelAli/CommandAPI/pull/395#discussion_r1063896585) works like this, so that's what I would suggest, but other solutions are possible. The key is to remember how the previous arguments were unpacked by not resetting the `arguments` list each loop iteration.

Overall, this is similar to what happened with #301 (If you want extra reading).